dissect/database/sqlite3/wal.py

```diff
@@ -0,0 +1,196 @@
+            name = getattr(fh, "name", None)
+            path = Path(name) if name else None
+        else:
+            if not isinstance(fh, Path):
+                fh = Path(fh)
+            path = fh
+            fh = path.open("rb")
+
+        self.fh = fh
+        self.path = path
+        self.header = c_sqlite3.wal_header(fh)
+
+        if self.header.magic not in WAL_HEADER_MAGIC:
+            raise InvalidDatabase("Invalid header magic")
+
+        self.checksum_endian = "<" if self.header.magic == WAL_HEADER_MAGIC_LE else ">"
+        self._checkpoints = None
+
+        self.frame = lru_cache(1024)(self.frame)
+
+    def frame(self, frame_idx: int) -> Frame:
+        frame_size = len(c_sqlite3.wal_frame) + self.header.page_size
+        offset = len(c_sqlite3.wal_header) + frame_idx * frame_size
+        return Frame(self, offset)
+
+    def frames(self) -> Iterator[Frame]:
+        frame_idx = 0
+        while True:
+            try:
+                yield self.frame(frame_idx)
+                frame_idx += 1
+            except EOFError:  # noqa: PERF203
+                break
+
+    @cached_property
+    def commits(self) -> list[Commit]:
+        """Return all commits in the WAL file.
+
+        Commits are frames where ``header.page_count`` specifies the size of the
+        database file in pages after the commit. For all other frames it is 0.
+
+        References:
+            - https://sqlite.org/fileformat2.html#wal_file_format
+        """
+        commits = []
+        frames = []
+
+        for frame in self.frames():
+            frames.append(frame)
+
+            # A commit record has a page_count header greater than zero
+            if frame.page_count > 0:
+                commits.append(Commit(self, frames))
+                frames = []
+
+                # TODO There might be data stored in later frames after the commit?
+
+        return commits
+
+    @cached_property
+    def checkpoints(self) -> list[Checkpoint]:
+        """Return deduplicated WAL commits (checkpoints), newest first.
+
+        Deduplicate commits by the salt1 value of their first frame. Later
+        commits overwrite earlier ones so the returned list contains the most
+        recent commit for each salt1, sorted descending.
+
+        References:
+            - https://sqlite.org/fileformat2.html#wal_file_format
+            - https://sqlite.org/wal.html#checkpointing
+        """
+        checkpoints_map: dict[int, Checkpoint] = {}
+        for commit in self.commits:
+            if not commit.frames:
+                continue
+            salt1 = commit.frames[0].header.salt1
+            # Keep the most recent commit for each salt1 (later commits overwrite).
+            checkpoints_map[salt1] = commit
+
+        return sorted(
+            checkpoints_map.values(),
+            key=lambda c: c.frames[0].header.salt1,
+            # TODO Should this be reverse? Reverse means: cp0 is the latest, cpN the oldest
```

Comment:
Reverse sorting is convenient for reading, but it's currently not consistent with `.commits` and `.frames`. Maybe don't sort it reverse?